src/irmin-pack/unix/lower_intf.ml

```diff
@@ -0,0 +1,98 @@
+(*
+ * Copyright (c) 2023 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open! Import
+
+module type Volume = sig
+  module Io : Io.S
+  module Errs : Io_errors.S
+
+  type t
+
+  type open_error =
+    [ Io.open_error
+    | `Closed
+    | `Double_close
+    | `Corrupted_control_file
+    | `Unknown_major_pack_version of string ]
```

Comment:
(k rambling) Cool! Could this mean that we're envisioning having a precondition/postcondition code style sometimes in the future? Or this is it still SF?

Reply:
I think that would be a bit too much :P .. and not really reasonable in presence of external side-effects (we can't freeze the filesystem to ensure that our preconditions stay valid, so the real operations can still crash later)
In the short term, I'm more concerned that the `error` polymorphic variant doesn't become a catch all if it starts bleeding internal error cases that don't make sense to the caller (I think it's better to `assert false` locally where we can easily guarantee that a `Double_close` is impossible, or produce a more semantic error `Volume_open_internal_error of [Closed|...]` to ease the caller error handling understanding)

Reply:
I'd love to talk more about this because I feel very similarly. I did have a pass of the code that tried to filter out errors that "didn't make sense" but the code ended up looking not great. So, I kept our current pattern of just passing through every error... 

If we came up with a consistent way to filter out errors that are programmer error, that would be :100:, and we could start applying it.